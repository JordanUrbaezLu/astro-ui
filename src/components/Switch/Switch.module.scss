@use "../../utility/globals";

.switchContainer {
  align-items: center;
  background: none;
  border: 0;
  cursor: pointer;
  display: flex;
  margin: 0;
  outline: none;
  padding: 0;
}

.switchPill {
  background: gray;
  border-radius: 999px;
  height: 22px;
  position: relative;
  transition: all 0.15s ease-in;
  width: 42px;
  z-index: 1;

  &.on {
    background: globals.$secondary-border-color-hover-and-focus;
  }
}

.switchIndicator {
  background: white;
  border-radius: 999px;
  content: "";
  display: inline-block;
  height: 16px;
  position: absolute;
  right: 23px;
  top: 3px;
  transition: all 0.15s ease-in;
  width: 16px;
  z-index: 3;

  &.on {
    background-color: globals.$secondary-background-color-active;
    transform: translateX(20px);
  }
}

.switchRipple {
  background: transparent;
  border-radius: 999px;
  content: "";
  display: inline-block;
  height: 30px;
  overflow: hidden;
  position: absolute;
  right: 16px;
  top: -4px;
  transition: all 0.15s ease-in;
  width: 30px;
  z-index: 4;

  &.on {
    transform: translateX(20px);
  }
}

.switchLabel {
  color: gray !important;
  margin-left: 8px;
  transition: all 0.15s ease-in;

  &.on {
    color: globals.$secondary-text-color-hover-and-focus !important;
  }
}
