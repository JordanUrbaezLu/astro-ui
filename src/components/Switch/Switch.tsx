import classNames from "classnames";
import * as React from "react";
import { Ripple } from "../Ripple";
import { Typography } from "../Typography/Typography";
import styles from "./Switch.module.scss";

export interface SwitchProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * If the Switch is on
   *
   * @default false
   */
  isOn?: boolean;
  /**
   * The label for the Switch
   */
  label?: string;
}

/**
 * @public
 *
 * @description
 *
 * The Switch component is used to toggle states.
 */
export const Switch: React.FC<SwitchProps> = (props) => {
  const { className, isOn = false, label, ...rest } = props;

  return (
    <button
      aria-checked={isOn}
      className={classNames(
        className,
        styles.switchContainer,
        isOn && styles.on
      )}
      role="switch"
      type="button"
      {...rest}
    >
      <span
        className={classNames(styles.switchPill, isOn && styles.on)}
      >
        <span
          className={classNames(
            styles.switchIndicator,
            isOn && styles.on
          )}
        />
        <span
          className={classNames(
            styles.switchRipple,
            isOn && styles.on
          )}
        >
          <Ripple duration={800} position="center" />
        </span>
      </span>
      {label && (
        <Typography
          className={classNames(
            styles.switchLabel,
            isOn && styles.on
          )}
        >
          {label}
        </Typography>
      )}
    </button>
  );
};

Switch.displayName = "Switch";
