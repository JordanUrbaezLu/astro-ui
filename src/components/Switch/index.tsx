/* istanbul ignore file */
export * from "./Switch";
