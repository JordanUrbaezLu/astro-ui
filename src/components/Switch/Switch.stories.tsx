import { Story } from "@storybook/react";
import * as React from "react";
import { Switch, SwitchProps } from "./Switch";
import { within, userEvent } from "@storybook/testing-library";
import { allowLayoutCalculations } from "../../../.storybook/utilities/allowLayoutCalculations";

export default {
  component: Switch,
  parameters: {
    chromatic: { delay: 2000 },
  },
  title: "Components/Inputs/Switch",
};

const Template: Story<Omit<SwitchProps, "isOn" | "onClick">> = (
  args
) => {
  const [switchState, toggleSwitchState] =
    React.useState<boolean>(false);

  return (
    <Switch
      {...args}
      isOn={switchState}
      onClick={() => toggleSwitchState(!switchState)}
    />
  );
};

export const Default = Template.bind({});

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: " Switch Label",
};
WithLabel.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("switch"))
  );
};
