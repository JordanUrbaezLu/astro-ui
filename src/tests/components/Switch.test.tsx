import { render, screen } from "@testing-library/react";
import { Switch } from "../../components/Switch/Switch";
import { axe } from "jest-axe";

describe("Switch", () => {
  test("Should render component correctly.", () => {
    // Fixes some jest-axe issues
    const { getComputedStyle } = window;
    window.getComputedStyle = (elt) => getComputedStyle(elt);
    // Fixes some jest-axe issues
    const { container } = render(
      <Switch isOn onClick={jest.fn()} label=" Switch Label" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Switch isOn onClick={jest.fn()} label="Switch Label" />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render label correctly.", () => {
    const label = "Switch Label";

    render(<Switch isOn onClick={jest.fn()} label={label} />);
    expect(screen.getByRole("switch")).toHaveTextContent(label);
  });

  test("Should render on correctly.", () => {
    render(<Switch isOn onClick={jest.fn()} label="Switch Label" />);
    expect(screen.getByRole("switch")).toHaveClass("on");
  });

  test("Should render off correctly.", () => {
    render(<Switch onClick={jest.fn()} label="Switch Label" />);
    expect(screen.getByRole("switch")).not.toHaveClass("on");
  });
});
