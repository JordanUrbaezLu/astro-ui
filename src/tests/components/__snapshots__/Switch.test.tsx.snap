// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Should render component correctly. 1`] = `
<button
  aria-checked="true"
  class="switchContainer on"
  role="switch"
  type="button"
>
  <span
    class="switchPill on"
  >
    <span
      class="switchIndicator on"
    />
    <span
      class="switchRipple on"
    >
      <div
        class="container"
        role="presentation"
      />
    </span>
  </span>
  <div
    class="switchLabel on typography typographyh2 primary normal"
  >
     Switch Label
  </div>
</button>
`;
